### PR TITLE
Resolve #63 seq2seq_attn example polish & Resolve #55 AttentionRNNDecoder interface refactor

### LINKED ISSUE #55: AttentionRNNDecoder interface refactor
Requiring `input_size` and `encoder_output_size`, especially in `AttentionRNNDecoder`, is too much, e.g., [here](https://github.com/asyml/texar-pytorch/blob/master/examples/seq2seq_attn/seq2seq_attn.py#L66-L69):
```
tx.modules.AttentionRNNDecoder(
            encoder_output_size=(self.encoder.cell_fw.hidden_size +
                                 self.encoder.cell_bw.hidden_size),
            input_size=self.target_embedder.dim + config_model.decoder
            ['attention']['attention_layer_size'],
            ...)
```

Can we refactor the interface a bit so that sth like
```
encoder_output_size=self.encoder.output_size,
input_size=self.target_embedder.dim
```
is enough?

related issue: https://github.com/asyml/texar-pytorch/issues/42

@gpengzhi @huzecong @TomNong

### LINKED ISSUE #63: seq2seq_attn example polish
- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L23 ---> `from texar.core import ...` : do not expose too many library internal details 

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L85 ---> `helper_train = self.decoder.create_helper(decoding_strategy="train_greedy")`: add readibility

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L107
Is it necessary to convert to `tensor`. Can this be done internally?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L111
Concat right after `enc_outputs` is obtained

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L113 
This should be in config file

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L128
`parameters()` vs `trainable_variables()` ?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L132-L133
it's weird to get `iterator` from `data_iterator`

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L137
Is it common?

- [x] https://github.com/asyml/texar-pytorch/blob/dc51de295da9e861d739bf40ee4902cf8669ca53/examples/seq2seq_attn/seq2seq_attn.py#L155
Mode in `{train, eval, predict}` or `{train, val}`

- [x] Add beam-search

### PR DESCRIPTION
**Pull request by gpengzhi**
#63 
1. add `beam_search` to `attn_decoder`
2. code refactor on `optimization`
3. polish `seq2seq_attn` exmaple
4. resolve #55

### CODE REVIEW

File: examples/seq2seq_attn/seq2seq_attn.py [hunk 1]
```diff
@@ -101,40 +106,39 @@ def forward(self, batch, mode):
...
+    test_data = tx.data.PairedTextData(hparams=config_data.test,
+                                       device=device)
     data_iterator = tx.data.TrainTestDataIterator(
         train=train_data, val=val_data, test=test_data)
 
     model = Seq2SeqAttn(train_data)
-    optimizer = get_optimizer(model.parameters(), config_model.opt)
-    train_op = get_train_op(optimizer, config_model.opt)
+    model.to(device)
+    train_op = optimization.get_train_op(params=model.parameters(),
```
Comment (huzecong) [inline examples/seq2seq_attn/seq2seq_attn.py]: I believe the intended usage suggested by @ZhitingHu was `tx.core.get_train_op`. Basically we only expose the first layer of packages to users.

File: texar/core/optimization.py [hunk 2]
```diff
@@ -289,15 +294,22 @@ def get_train_op(optimizer: Optimizer,
...
+                         "simultaneously.")
+
+    if optimizer is None and params is not None:
+        optimizer = get_optimizer(params, hparams)
+    if optimizer is not None:
+        scheduler = get_scheduler(optimizer, hparams)
     grad_clip_fn = get_grad_clip_fn(hparams)
 
-    params_list = []
+    params_list: List = []
```
Comment (huzecong) [inline texar/core/optimization.py]: Should the type be `List[nn.Parameter]`?

File: texar/modules/decoders/rnn_decoders.py [hunk 3]
```diff
@@ -660,6 +723,55 @@ def forward(  # type: ignore
...
+
+        def _prepare_beam_search(x):
+            x = torch.unsqueeze(x, dim=1)
+            tile_dims = [1] * x.dim()
+            tile_dims[1] = beam_width
+            x = x.repeat(tile_dims)
+            shape = list(x.size())
+            shape[0] *= shape[1]
+            shape.pop(1)
+            x = x.view(shape)
```
Comment (huzecong) [inline texar/modules/decoders/rnn_decoders.py]: I think this entire function can be simplified as:
```python
x = x.unsqueeze(1).repeat(-1, beam_width, *([-1] * (x.dim() - 1)))
x = x.view(-1, *x.size()[2:])
```

File: texar/modules/decoders/rnn_decoders.py [hunk 4]
```diff
@@ -660,6 +723,55 @@ def forward(  # type: ignore
...
+        def _symbols_to_logits_fn(ids, state):
+            ids = ids[:, -1]
+            inputs = embedding_fn(ids)
+            wrapper_outputs, wrapper_state = self._cell(
+                inputs, state, memory_beam_search,
+                memory_sequence_length_beam_search)
+            logits = self._output_layer(wrapper_outputs)
+            return logits, wrapper_state
+
+        outputs, log_prob = beam_search(  # type: ignore
```
Comment (huzecong) [inline texar/modules/decoders/rnn_decoders.py]: Why do we need a `# type: ignore` here?

File: texar/utils/beam_search.py [hunk 5]
```diff
@@ -36,8 +36,10 @@
 INF = 1.0 * 1e7
 
 
-def gather_nd(params: torch.Tensor, indices: torch.Tensor) -> torch.Tensor:
-
+def gather_nd(params: Union[torch.Tensor, int],
```
Comment (huzecong) [inline texar/utils/beam_search.py]: Why would this and following methods take `int` as an argument?

File: examples/seq2seq_attn/seq2seq_attn.py (hunk 1 above)
Comment (ZhitingHu) [reply examples/seq2seq_attn/seq2seq_attn.py]: Yes, let's call it by `tx.core.get_train_op` as in the Texar TF examples

File: examples/seq2seq_attn/seq2seq_attn.py [hunk 6]
```diff
@@ -101,40 +106,39 @@ def forward(self, batch, mode):
             start_tokens = torch.ones_like(batch['target_length']) * \
```
Comment (ZhitingHu) [inline examples/seq2seq_attn/seq2seq_attn.py]: Use `new_full` to create start_tokens, pls refer to PR https://github.com/asyml/texar-pytorch/pull/76

File: texar/utils/beam_search.py (hunk 5 above)
Comment (gpengzhi) [reply texar/utils/beam_search.py]: `state` used in `beam_search` is `AttentionWrapperState`. We here have `int` because of `time` in  `AttentionWrapperState`. I deleted `time` to simplify the interface in `beam_search`.

File: texar/modules/decoders/rnn_decoders.py (hunk 4 above)
Comment (gpengzhi) [reply texar/modules/decoders/rnn_decoders.py]: `self._vocab_size` could be `None` (https://github.com/asyml/texar-pytorch/blob/master/texar/modules/decoders/decoder_base.py#L92).

File: texar/utils/beam_search.py (hunk 5 above)
Comment (huzecong) [reply texar/utils/beam_search.py]: I see. Yeah `time` is probably unnecessary, and I kind of think that some other fields in `AttentionWrapperState` are unnecessary but I'm not sure.

File: texar/modules/decoders/rnn_decoders.py (hunk 4 above)
Comment (huzecong) [reply texar/modules/decoders/rnn_decoders.py]: In general it's much better to use assertions to guide the type checker. This adds almost no overhead anyway.

File: texar/utils/beam_search.py (hunk 5 above)
Comment (ZhitingHu) [reply texar/utils/beam_search.py]: `AttentionWrapperState` in TF [here](https://www.tensorflow.org/api_docs/python/tf/contrib/seq2seq/AttentionWrapperState) has `time`. (Is `time` used in TF?)  I'd suggest keeping it the same as the TF interface even if it's currently not used (it might be used in other attention mechanisms, probably in the future)

File: texar/core/optimization.py [hunk 7]
```diff
@@ -271,12 +272,17 @@ def get_grad_clip_fn(hparams: Optional[Union[HParams,
     return grad_clip_fn
 
 
-def get_train_op(optimizer: Optimizer,
+def get_train_op(params: Optional[Union[List[torch.Tensor],
```
Comment (ZhitingHu) [inline texar/core/optimization.py]: [Minor issue] The type of `params` looks very complex. Is it necessary or can it be simplified? (ok if not)

File: texar/core/optimization.py [hunk 8]
```diff
@@ -271,12 +272,17 @@ def get_grad_clip_fn(hparams: Optional[Union[HParams,
...
+                                        List[Dict[str,
+                                                  List[torch.Tensor]]]]] = None,
+                 optimizer: Optional[Optimizer] = None,
```
Comment (ZhitingHu) [inline texar/core/optimization.py]: `get_train_op` should also have an argument `scheduler=None` to allow users to pass in any customized scheduler, like optimizer.

File: texar/utils/beam_search.py (hunk 5 above)
Comment (gpengzhi) [reply texar/utils/beam_search.py]: If I remember correctly, in TF `time` is only used to keep track of the time step in the decoding procedure. If we really want to keep it the same as the TF interface, we could either make modifications in `beam_search` or `utils.map_structure.` (I probably can follow the similar way of `no_map_type` to bypass this issue).

File: examples/seq2seq_attn/seq2seq_attn.py [hunk 9]
```diff
@@ -72,20 +73,18 @@ def __init__(self, train_data):
```
Comment (ZhitingHu) [inline examples/seq2seq_attn/seq2seq_attn.py]:
```
tx.modules.AttentionRNNDecoder(
  ...
  input_size=self.target_embedder.dim + config_model.decoder
              ['attention']['attention_layer_size'],
```
`input_size` is not fixed yet. Any problem here?

File: texar/utils/beam_search.py (hunk 5 above)
Comment (ZhitingHu) [reply texar/utils/beam_search.py]: Yes, let's keep `time` as in TF. It's likely to be useful in the future.

File: examples/seq2seq_attn/seq2seq_attn.py (hunk 9 above)
Comment (gpengzhi) [reply examples/seq2seq_attn/seq2seq_attn.py]: Sorry for the late reply. I will fix #55 in this pull request as well.

Comment (ZhitingHu) [reply examples/seq2seq_attn/seq2seq_attn.py]: Since the interface is changed, does the `AttentionRNNDecoder` doc (about `input_size`) need to be updated as well?
